src/System.Management.Automation/engine/MshObject.cs

```diff
@@ -1122,18 +1122,24 @@ internal static string ToStringEnumerator(ExecutionContext context, IEnumerator
             while (enumerator.MoveNext())
             {
                 object obj = enumerator.Current;
-                returnValue.Append(PSObject.ToString(context, obj, separator, format, formatProvider, false, false));
-                returnValue.Append(separatorToUse);
+                if (TryFastTrackPrimitiveTypes(obj, format, formatProvider, out string objString))
+                {
+                    returnValue.Append(objString);
+                }
+                else
+                {
+                    PSObject mshObj = PSObject.AsPSObject(obj);
+                    returnValue.Append(PSObject.ToString(context, mshObj, separator, format, formatProvider, false, false));
+                    returnValue.Append(separatorToUse);
+                }
```

Comment:
`DateTime` has a higher chance to get a custom `ToString` definition in ETS.
I think we should just strictly restrict the special handling to `string` and true primitive types (namely `type.IsPrimitive` is true).
```c#
var objType = obj.GetType();
if (objType != typeof(string) && !objType.IsPrimitive)
{
    obj = PSObject.AsPSObject(obj);
}

returnValue.Append(PSObject.ToString(context, obj, separator, format, formatProvider, false, false));
returnValue.Append(separatorToUse);
```

Reply:
1. What about Decimal?
2. The PR intention is performance. In your sample we still do some duplication type checks. I'd refactor TryFastTrackPrimitiveTypes() in my sample with using IsPrimitive()

Reply:
In my mind, a perf change shouldn't result in any behavioral changes, and that's why I'm concerned about you being aggressive here.
I would much prefer to be conservative
- Only treat string specially and keep on wrapping other values into `PSObject` **if we go for the most conservative manner**;
- Only treat `IsPrimitive` objects specially and keep wrapping other values **if we go a bit less conservative**

As for `decimal`, it's OK that we don't cover it. Again, it's not a functionality change that we should cover all scenarios, but a perf change that targets improve common scenarios.
As for `IntPtr` being a primitive type, I think it's fine too, as it's not really possible to use `IntPtr` directly in PowerShell script anyways.

My personal preference: only treat string specially. /cc @JamesWTruher @rjmholt what's your thoughts on this?

Reply:
>  a perf change shouldn't result in any behavioral changes, and that's why I'm concerned about you being aggressive here

I am full agree and very appreciate your help!
The "aggressive" come from original issue where we consider primitive types too. I agree that for join operator strings are the most expected types for the operator. Nevertheless, I believe that there are many scenarios (like custom output formatting) when users will use the join operator in loops with primitive types. Notice that we change only an intermediate code and we added tests to exclude breaking change. So I believe we could be less conservative here. But I do not found how test the enumerator change and I'd revert the change.

Reply:
> /cc @JamesWTruher @rjmholt what's your thoughts on this?

I think specialising for `string` only makes the most sense. It's by far the most common, and the place where we don't break things. I don't think there are many people out there doing lots of string joins on other primitive types, but by breaking ETS we'd lose functionality